keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,582 @@
+
+![alt text](./mem_tables_runtime.png)
+
+However, if **Out-of-Memory** (OOM) situation occurs, i.e., a particular memory limit is reached and crossed (e.g. by the containers) within _K8S_ node, the node triggers one of [the mechanisms](#kubernetes-nodes-memory-management-mechanisms-and-their-relation-to-the-memory-manager). Depending on what a particular limit is crossed (`cgroups` memory limit, `hard-eviction-treshold` memory limit, etc.), a suitable mechanism becomes triggered accordingly. As already noticed, such an OOM situation is handled by *kubelet* or system (OOM kernel killer) and not by *Memory Manager* itself. 
+
+### Simulation - how the Memory Manager works? (by examples)
+
+In this section, a number of slides is intended to illustrate the operation of the Memory Manager in various situations, i.e., to show how the pods in different QoS classes (Guaranteed, BestEffort/Burstable) are handled by the Memory Manager or other means, and how the disjoint groups of NUMA nodes are dynamically managed (created or removed) by the Memory Manager.  
+
+![alt text](./mm_simulation.jpg)
+
+### Hints Generation for Topology Manager
+
+In brief, the Memory Manager calculates topology hints for a container in Guaranteed QoS Pod for conventional memory and hugepages of all sizes. A topology hint is also known as _NUMA affinity_ which represents a possible set of NUMA nodes that has enough capacity to satisfy container's memory demand for all demanded memory types.
+
+The concept of hint generation that ensures the guaranteed memory allocation was explained [here][sig-node-memory-manager-presentation] (slides 17-20). The presentation illustrates on how hint generation is performed by the Memory Manager.
+
+<!--
+To generate a NUMA affinity for a container, Memory Manager refers to `Memory Map` and decides if _Free_ memory of _Allocatable Zone_ for all memory types from each NUMA node is enough to assign for the container. After collecting _NUMA affinity_ for each NUMA node from the perspective of memory, Memory Manager merges them to a topology hint and send it back to Topology Manager.
+
+- Here is a brief example for two socket machine.
+  - A Pod requests 2Gi of memory and 1 hugepages-1Gi.
+  - Memory Manager found that NUMA node #0 has enough memory and hugepage but NUMA node #1 does not so that the calculated NUMA affinity is NUMA node #0.
+  - Memory Manager returns the result as a topology hint with NUMA node #0.
+-->
+
+### New Flags and Configuration of the Memory Manager
+
+#### Feature Gate Flag 
+
+A new feature gate flag will be added to enable the Memory Manager feature. This feature gate will be disabled by default in the initial releases.
+
+Syntax:
+`--feature-gate=MemoryManager=false|true`
+
+#### Memory Manager Policy Flag  
+
+The `static` value of the flag enables the guaranteed memory allocation, whereas `none` disables it. 
+
+Syntax:
+`--memory-manager-policy=static|none`
+
+#### Reserved Memory Flag
```

Comment:
This section is a little hard to follow with the way it is worded. I believe the main points are:
1) Node allocatable is already a feature of the kubelet
2) A set of flags exist to express the total amount of memory reserved for the system so that the "allocatable"  memory of the system can be calculated.
3) These flags are `kube-reserved`, `system-reserved`, and `eviction-threshold`.
3) The sum of the memory expressed by these flags declares the total *amount* of memory reserved for the system
4) Your new flag, `--reserved-memory`, then allows one to declare *how* this total reserved memory should be split across NUMA nodes.

Assuming what I've said is correct, I think a flow from top to bottom, introducing the flag this way will make it clearer as to its intended purpose.

Reply:
Would it make more sense to align with how "reserved-cpus" works?  That is, you can only use the "--reserved-cpus" options if CPUs are not reserved via  kube-reserved or system-reserved.

That way you would only need to specify the reserved memory in one place rather than needing to ensure that two options agree with each other.

Also, do we want to consider making the memory eviction threshold NUMA-aware as well?  Is it possible for us to get into OOM trouble if one NUMA node is out of memory and the other isn't?

Reply:
@cbf123



> That way you would only need to specify the reserved memory in one place rather than needing to ensure that two options agree with each other.

The problem that `kube-reserved` and `system-reserved`, not NUMA aware, so we can not rely on these options during the calculation of the Allocatable memory.



> Also, do we want to consider making the memory eviction threshold NUMA-aware as well? Is it possible for us to get into OOM trouble if one NUMA node is out of memory and the other isn't?

We pin the memory and calculate the free memory only for guaranteed containers, so it possible that the best-effort or burstable container will have OOM issue(because they have bigger OOM score), but I think the standard eviction-threshold mechanism should be enough in this case because we do not have memory pinning for such containers and it should have OOM issues only when the total amount of memory on the node is dropped under some boundary.

Reply:
@klueska your flow was followed, which resulted in a rewording:

> [Node Allocatable Feature](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/) is already a feature of K8S, and it is commonly used by node administrators to reserve resources in a K8S node (for the kubelet processes or system processes) in order to enhance the stability of K8S nodes. A group of flags is used to specify the total amount of reserved memory at a K8S node. This value is then used to calculate the real offered amount of node's "allocatable" memory. Also, K8S scheduler uses "allocatable" to improve the scheduling. The foregoing flags are `kube-reserved`, `system-reserved` and `eviction-threshold`. Their sum accounts for the total amount of reserved memory.    
> 
> The new `--reserved-memory` was added to allow for this total reserved memory to be split (by a node administrator) across many NUMA nodes.